pydm/data_plugins/__init__.py

```diff
@@ -96,31 +101,167 @@ def plugin_for_address(address):
 
-def load_plugins_from_path(locations, token):
+    Returns
+    -------
+    plugins : list of PyDMPlugin classes
+        The plugin classes.
+    """
+    module = import_module_by_filename(source_filename)
+    return list(
+        set(
+            obj
+            for _, obj in inspect.getmembers(module)
+            if _is_valid_plugin_class(obj)
+        )
+    )
+
+
+def find_plugins_from_path(
+    path: str, token: str = config.DATA_PLUGIN_SUFFIX
+) -> Generator[Type[PyDMPlugin], None, None]:
+    """
+    Yield all data plugins found in the provided path.
+
+    Parameters
+    ----------
+    path : str
+        The path to look for plugins.
+    token : str, optional
+        The suffix that plugin files are expected to have.
+    """
+
+    for root, _, files in os.walk(path):
+        if root.split(os.path.sep)[-1].startswith("__"):
+            continue
+
+        logger.debug("Looking for PyDM Data Plugins at: %s", root)
+        for name in files:
+            if name.endswith(token):
+                yield from _get_plugins_from_source(os.path.join(root, name))
+
+
+def find_plugins_from_entrypoints(
+    key: str = config.ENTRYPOINT_DATA_PLUGIN,
+) -> Generator[Type[PyDMPlugin], None, None]:
+    """
+    Yield all PyDMPlugin classes specified by entrypoints.
+
+    Uses ``entrypoints`` to find packaged external tools in packages that
+    configure the ``pydm.data_plugin`` entrypoint.
+
+    Parameters
+    ----------
+    key : str, optional
+        The entrypoint key.
+    """
+    for entry in entrypoints.get_group_all(key):
+        logger.debug("Found data plugin entrypoint: %s", entry.name)
+        try:
+            plugin_cls = entry.load()
+        except Exception as ex:
+            logger.exception(
+                "Failed to load %s entry %s: %s",
+                key, entry.name, ex
+            )
+            continue
+
+        if not _is_valid_plugin_class(plugin_cls):
+            logger.warning(
+                "Invalid plugin class specified in entrypoint "
+                "%s: %s",
+                entry.name, plugin_cls
+            )
+            continue
+
+        yield plugin_cls
+
+
+def _is_valid_plugin_class(obj: Any) -> bool:
+    """Is the object a data plugin class?"""
+    return (
+        inspect.isclass(obj)
+        and issubclass(obj, PyDMPlugin)
+        and obj is not PyDMPlugin
+    )
+
+
+def load_plugins_from_entrypoints(
+    key: str = config.ENTRYPOINT_DATA_PLUGIN
+) -> Dict[str, PyDMPlugin]:
     """
     Load plugins from file locations that match a specific token
 
+    Parameters
+    ----------
+    key : str, optional
+        The entrypoint key.
+
+    Returns
+    -------
+    plugins : dict
+        Dictionary of plugins add from this folder
+    """
+    added_plugins = dict()
+    for plugin in find_plugins_from_entrypoints(key):
+        if plugin.protocol is not None:
+            added_plugins[plugin.protocol] = add_plugin(plugin)
```

Comment:
This seems like a good point for a warning/error message, a data plugin probably intends to select a protocol

Reply:
Resolved in 52849a12ac00a0a4d6afafb2141ec7581d8ea5ae